### DISCUSSION TITLE
Proper way of versioning a stack of dependencies

### DISCUSSION THREAD

**Discussion by bartosik-hep** [spack/spack#36284] [Q&A]
We use Spack for building an extensive list of software, which is supposed to be used in several different environments:
* virtual machines of a computing cluster with shared storage *(using upstream installations)*
* private machines
* Docker containers

To have everything defined in one place we've create a [mucoll-stack](https://github.com/MuonColliderSoft/mucoll-spack/blob/master/packages/mucoll-stack/package.py) package, representing the whole software stack, with all the necessary software added as dependencies.

As different dependencies evolve with time, we want to periodically issue a release of the whole stack, where versions and variants of all the dependencies are explicitly defined, like when you concretize the `mucoll-stack` package. 

**My question is:** what is the optimal way of versioning the `mucoll-stack` package such that each version corresponds to a specific combination of specs of all its dependencies?

The only relevant piece I found in [Spack documentation](https://spack.readthedocs.io/en/latest/build_settings.html) is about package settings file `packages.yaml`, but it is meant to be used for the whole environment, while we need the versions and variants to be applied across different environments depending on the version of the `mucoll-stack` package.

I hope there is an elegant solution that I'm overlooking...

**Comment by psakievich** [comment]
You can use [conditional dependencies](https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-dependencies) to do things like `depends_on(foo_depency@1.1, when='@3.0')` for your `mucoll-stack` package.  This would allow you to manage the entire stack through the package and control what gets built by the version of the `mucoll-stack` `BundlePackage`.
